### DISCUSSION TITLE
EPMacro does not compile when building E+ from sources for ARMv8 architecture (RPi-4B) inside Docker

### DISCUSSION THREAD

**Discussion by juancaoviedo** [NREL/EnergyPlus#8876] [Q&A]
Hello,

I want to compile E+ 9.x.0 for ARMv8 architecture (RPI-4B) using Docker. However, when using:
```
RUN cmake /EnergyPlus-9.x.0/ 
RUN make -j4 install 
```
the compiler does not create EPMacro and ExpandObjects in the destiny directory (/usr/local in my case). 
I found ExpandObjects CMakeLists.txt inside /EnergyPlus-9.x.0/src/ExpandObjects. So, by running
```
RUN cmake ExpandObjects 
RUN make -j4 install 
```
the problem was solved (I think; I did not test yet but at least the sequence of commands creates the ExpandObjects file in the destiny directory). However, I can not find the EPMacro CMakeLists.txt inside the /EnergyPlus-9.x.0 directory to compile it. The only thing I found was /EnergyPlus-9.x.0/bin/EPMacro, which contains three directories: Linux, Windows, Mac. Each of the directories contains an EPMacro executable. However, I believe these EPMacro files are precompiled, and copying them to the destiny directory will not work since they are precompiled for different architectures, right? (I tried to copy them, it didn't work).

I read in the [build E+](https://github.com/NREL/EnergyPlus/wiki/BuildingEnergyPlus) page that it is possible to use ccmake to choose the characteristics you want to E+ to compile. However, since I'm using a Dockerfile to create an image, I can not use interactive selection as suggested on the wiki page. And, honestly, I do not know if that will solve my problem. At this point, I do not know what to do. 

So, I want to know if it is possible to compile EPMacro from the sources for ARMv8 architecture. I will highly appreciate some light on this subject and thank you very much in advance. 

Pdt: I'm doing the process because I want to use E+ with PyFMI, which requires EPMacro and ExpandObjects to function properly. Thank you again!

**Comment by Myoldmopar** [comment]
About EPMacro compilation: We are not able to compile EPMacro anymore.  I believe the Intel Fortran compiler can still do it, but we do not compiler with Intel.  I don't recall exactly what the problem was, maybe corner-case Hollerith constant support?  Anyway, it doesn't matter much, the point is that we have been packaging up precompiled EPMacro binaries for quite some time.

Now to your problem, the fact that you aren't getting ExpandObjects or EPMacro makes me think you have BUILD_FORTRAN turned off.  When you run CMake, pass some additional arguments: `cmake -DBUILD_FORTRAN=ON -DBUILD_PACKAGE=ON /EnergyPlus-9.x.0`.  Or something like that.  That will be the equivalent of running ccmake and dynamically enabling those cmake variables.  You can mimic some of your Docker build process off of our Github Action package builds: https://github.com/NREL/EnergyPlus/blob/develop/.github/workflows/linux_release.yml#L52

> I want to use E+ with PyFMI, which requires EPMacro and ExpandObjects to function properly

What?  While I'll concede that you may want to run ExpandObjects on some HVAC Template stuff, I cannot imagine you actually need EPMacro.  ExpandObjects should build on your platform once you turn on those CMake variables.  EPMacro won't run since it is the wrong platform.  Probably just having the dummy EPMacro binary sitting there will suffice.

**Reply by juancaoviedo** [reply]
Thank you very much.

**Comment by juancaoviedo** [comment]
Thank you very much for your explanatory answer. I ran the CMakeFiles.txt by setting the options to ON as you suggested (I used sed):
```
RUN sed -i 's/option(BUILD_PACKAGE "Build package" OFF)/option(BUILD_PACKAGE "Build package" ON)/g' CMakeLists.txt
RUN sed -i 's/option(BUILD_TESTING "Build testing targets" OFF)/option(BUILD_TESTING "Build testing targets" ON)/g' CMakeLists.txt
RUN sed -i 's/option(BUILD_FORTRAN "Build Fortran stuff" OFF)/option(BUILD_FORTRAN "Build Fortran stuff" ON)/g' CMakeLists.txt
RUN cmake EnergyPlus-9.5.0
RUN make -j4 install
```

Now, the ExpandObjects and EPMacro are compiled in the destiny folder. However, when I run the PyFMI instruction, I still get the error with EPMacro:
```
Reading input and weather file for preprocessor program.
The IDF version of the input file ///tmp/JModelica.org/jm_tmpnpwiw55e//resources//f_312e95.idf starts with 9
Successfully finish reading weather file.
This is the Begin Month: 1
Time (1) set is smaller than minimun allowed (1 day). Day will be set to 1.
This is the Day of the Begin Month: 1
This is the End Month: 1
Time (1) set is smaller than minimun allowed (1 day). Day will be set to 1.
This is the Day of the End Month: 1
This is the New Day of Week: SUNDAY
Running EPMacro...
sh: 1: /usr/local/EPMacro: Exec format error
ExpandObjects Started.
ExpandObjects Finished with Error(s). Time:     0.000
EnergyPlus Starting
EnergyPlus, Version 9.5.0-Unknown (No OpenGL), YMD=2021.07.08 19:41
**FATAL:Errors occurred on processing input file. Preceding condition(s) cause termination.
EnergyPlus Run Time=00hr 00min  0.60sec
Program terminated: EnergyPlus Terminated--Error(s) Detected.
```

The EPMacro gives the "Exec format error" which I think is related to the problem of platform compatibility. Regarding the error of EnergyPlus, I don't think there is a problem with the IDF/FMU, since exactly the same version of the IDF/FMU runs in an AMD machine (with precisely the exact PyFMI instructions). 

For some reason, the PyFMI calls EPMacro and ExpandObjects when initializing the simulation, even though the IDF does not contain measures or HVAC systems. So, do you think there is something we can do to work around this error? There is a way to compile EPMacro for ARMv8? Should I try to deactivate the EPMacro call from PyFMI? Better to try with energysim instead of PyFMI? Better to change to AMD and not bother with the Raspberry Pi? 

It seems to me that E+ works for ARMv8 architecture (I launched a command line simulation and worked fine). Maybe the problem is just the EPMacro, the ExpandObjects seems to compile good (at leas I do not receive the Exec format error). 

Thank you very much again for your help,

**Comment by Myoldmopar** [comment]
Let me say it again: we don't compile EPMacro.  There is a difference between what we do for EPMacro and ExpandObjects.  ExpandObjects is compiled from source into a binary on the platform that it is built.  EPMacro is precompiled, assuming the platform ahead of time.

So this all makes sense.  You should definitely just disable the EPMacro call from PyFMI and it should work fine.
